### PR TITLE
Added support for regex and arrays in Express.Router.use

### PR DESCRIPTION
**Pull request by IOAyman**
Added support for regex and arrays in Express.Router.use
`Express.Router.use( path: RexExp | string | (RegExp | string)[], ....`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,4 @@ typings/
 node_modules/
 test/bundle.d.ts
 *.js
-
+.idea/
```
Comment (felixfbecker) [inline .gitignore]: I don't want any editor specific ignores in the repo. You can add this to your own `.git/info/exclude`.

File: index.d.ts [hunk 2]
```diff
@@ -1,3 +1,2 @@
-
```
Comment (felixfbecker) [inline index.d.ts:1]: Please exclude this from the PR

File: lib/router/index.d.ts [hunk 3]
```diff
@@ -278,6 +278,8 @@ declare namespace createRouter {
 
         use(...handlers: HandlerArgument[]): this;
         use(mountPoint: string, ...handlers: HandlerArgument[]): this;
+        use(mountPoint: RegExp, ...handlers: HandlerArgument[]): this;
+        use(mountPoint: (string | RegExp)[], ...handlers: HandlerArgument[]): this;
```
Comment (felixfbecker) [inline lib/router/index.d.ts]: Is there any reason these signatures can't be `mountPoint: string | RegExp | (string | RegExp)[]`?

Comment (blakeembrey) [inline lib/router/index.d.ts]: Can we just make this one union instead of many overloads? If possible, this would be good to re-use from the router too (perhaps we can make a `type Path`?).

File: test/app.ts [hunk 4]
```diff
@@ -1,13 +1,12 @@
-
 import express = require('express');
 import router from './router';
 
 const app: express.Application = express();
 
 // requesthandler
-const requestHandler: express.RequestHandler = (req, res, next) => {
+const requestHandler: express.RequestHandler = (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
Comment (felixfbecker) [inline test/app.ts]: Is this change needed? Does it compile without it? Support for type inference would be nice, that's why I added this line as it was

File: test/app.ts [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
     console.log(req.ip);
-    res.json({hello: 'world'});
+    res.json({ hello: 'world' });
```
Comment (felixfbecker) [inline test/app.ts]: Please keep stuff like this out of the PR

File: typings.json [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
   "homepage": "https://github.com/felixfbecker/typed-express",
-  "dependencies": {
-    "serve-static": "registry:npm/serve-static#1.11.1+20160810053450"
-  },
-  "version": "4.14.0",
   "globalDependencies": {
     "node": "registry:env/node#6.0.0+20160723033700"
+  },
+  "dependencies": {
+    "serve-static": "registry:npm/serve-static#1.11.1+20160810053450"
```
Comment (felixfbecker) [inline typings.json:6]: Please don't resort the JSON

**Comment by felixfbecker** [comment]
Added some comments :)

File: test/app.ts (hunk 4 above)
Comment (IOAyman) [inline test/app.ts]: I've made that change basically because you asked the compiler for `"noImplicitAny": true`

File: test/app.ts [hunk 7]
```diff
@@ -43,6 +42,8 @@ console.log(app.get('some_setting'));
 
 // mount router
 app.use('/mountpath', router);
+app.use(/\/mountpath/, router);
+app.use(['/mountpath', /\/mountpath/], router);
```
Comment (felixfbecker) [inline test/app.ts:47]: Please add one more test that checks a mixed array of strings and RegExps

File: test/app.ts (hunk 5 above)
Comment (IOAyman) [inline test/app.ts]: Will do.

File: typings.json (hunk 6 above)
Comment (IOAyman) [inline typings.json:6]: will do

File: .gitignore [hunk 8]
```diff
@@ -2,4 +2,3 @@ typings/
 node_modules/
 test/bundle.d.ts
 *.js
-
```
Comment (felixfbecker) [inline .gitignore:5]: editorconfig is configured with `insert_final_newline`. Please honor this or it will create further unneeded diffs in future commits ;)

Comment (IOAyman) [inline .gitignore:5]: I `git ckeckout felix/master .gitignore` that in previous commits :)

**Comment by IOAyman** [comment]
Thanks for the little notes :)  Are we good now?

File: .gitignore (hunk 8 above)
Comment (felixfbecker) [inline .gitignore:5]: Apparently not :D

File: test/app.ts (hunk 4 above)
Comment (felixfbecker) [inline test/app.ts]: Yeah but on master the type is not infered to `any` but properly to `express.Request` etc. Did your changes change that? If yes that would be a regression and breaking change.

**Comment by felixfbecker** [comment]
Almost :)

File: test/app.ts [hunk 9]
```diff
@@ -1,11 +1,10 @@
-
```
Comment (felixfbecker) [inline test/app.ts:1]: Add back this line please and we're ready to merge

File: test/app.ts (hunk 4 above)
Comment (IOAyman) [inline test/app.ts]: Ah I see! Sorry, my bad, you're right :)  Thanks!
I'll revert that line.
